### PR TITLE
2451 item merge (stock_line)

### LINKED ISSUE #2451: Item merging

### PR DESCRIPTION
**Pull request by Chris-Petty**
Fixes #2451 

# 👩🏻‍💻 What does this PR do? 

- 1.06.00 migration adding `item_link` and updating the `stock_on_hand` sql view.
- Migration removes `stock_line.item_id`, adds `stock_line.item_link_id`.
- Migration tests naturally.
- Added repository `item_link_row`. I got the functions in there compiling, but they aren't used yet so haven't tested. They'll be used in future PRs doing sync side of this.
- Fixed tests that relied on stock_line.item_id previously (they largely just needed `item_link` table to be populated correctly).
- `stock_line` repository adjusted to discretely use `item_link` to retrieve the related `item` (so API isn't adjusted, callers still use `stock_line.item_id`)

Things this does not touch which will be other issues:
- `invoice_line`
- `stock_take_line`
- `master_list_line`
- Syncing `item_link`
- Handling merge messages from the central server
- Anything else you can think of 😉

# 🧪 How has/should this change been tested? 

I've mainly just tested this in sql tests so far HA. Compiling is enou... There is a lot more I should do 🙂:
- Do all of the following both on sqlite and postgres
- Do a stocktake (snapshot quantity should be right)
- Can receive stock in a shipment
- Can send stock in a shipment 
- Stock on hand was shown and updated correctly when doing the above exercises

## 💌 Any notes for the reviewer?

Basically everything relating to `stock_line` should behave as if nothing changed.

## 📃 Documentation

_No user facing changes..._ _(hopefully)_

### CODE REVIEW

File: server/repository/src/migrations/v1_06_00/item_line_add_item_link_id.rs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+              SUM(pack_size * available_number_of_packs) AS available_stock_on_hand
+            FROM
+              store_items
+            WHERE
+              store_items.available_number_of_packs > 0
+            GROUP BY
+              item_id,
+              store_id
+          ) AS stock ON stock.item_id = items_and_stores.item_id
+          AND stock.store_id = items_and_stores.store_id;
```
Comment (Chris-Petty) [inline server/repository/src/migrations/v1_06_00/item_line_add_item_link_id.rs:68]: tangent: in vscode I installed a formatter and made this query in a different file

File: server/repository/src/db_diesel/stock_line_row.rs [hunk 2]
```diff
@@ -40,6 +40,7 @@ joinable!(stock_line -> barcode (barcode_id));
 #[table_name = "stock_line"]
 pub struct StockLineRow {
     pub id: String,
+    #[column_name = "item_link_id"]
```
Comment (Chris-Petty) [inline server/repository/src/db_diesel/stock_line_row.rs:43]: This is a very subtle but critical change that makes this PR work - the struct and repository API remains `item_id`, but the table schema actually stores the value in column `item_link_id`.

File: server/repository/src/db_diesel/stock_line.rs [hunk 3]
```diff
@@ -188,7 +201,7 @@ fn create_filtered_query(filter: Option<StockLineFilter>) -> BoxedStockLineQuery
         } = f;
 
         apply_equal_filter!(query, id, stock_line_dsl::id);
-        apply_equal_filter!(query, item_id, stock_line_dsl::item_id);
+        apply_equal_filter!(query, item_id, item::id);
```
Comment (Chris-Petty) [inline server/repository/src/db_diesel/stock_line.rs:204]: I'm not sure if this change is good haha. The alternative is `item_link::item_id` I think. Rust compiler should apply to my comments it'd help me be more confident of that statement.

File: server/repository/src/db_diesel/stock_line.rs [hunk 4]
```diff
@@ -160,15 +161,27 @@ impl<'a> StockLineRepository<'a> {
...
+allow_tables_to_appear_in_same_query!(stock_line, item_link);
+allow_tables_to_appear_in_same_query!(item, item_link);
+allow_tables_to_appear_in_same_query!(location, item_link);
+allow_tables_to_appear_in_same_query!(name, item_link);
+allow_tables_to_appear_in_same_query!(barcode, item_link);
+
 fn create_filtered_query(filter: Option<StockLineFilter>) -> BoxedStockLineQuery {
     let mut query = stock_line_dsl::stock_line
-        .inner_join(item_dsl::item)
+        .inner_join(item_link_dsl::item_link.inner_join(item_dsl::item))
```
Comment (Chris-Petty) [inline server/repository/src/db_diesel/stock_line.rs:184]: Instead of returning item as before, we join on item_link and get the item related to the item_link.item_id, returned as a tuple. Then we get a tuple of `(ItemLinkRow, ItemRow)` out of the query. In `to_domain()` we just ditch the link and return the item as if nothing weird happened.

File: server/repository/src/db_diesel/item.rs [hunk 5]
```diff
@@ -162,13 +162,13 @@ impl<'a> ItemRepository<'a> {
         //     diesel::debug_query::<DBType, _>(&final_query).to_string()
         // );
 
-        let result = final_query.load::<ItemAndMasterList>(&self.connection.connection)?;
+        let result = final_query.load::<ItemAndUnit>(&self.connection.connection)?;
```
Comment (roxy-dao) [inline server/repository/src/db_diesel/item.rs:165]: Nice catch~

File: server/repository/src/db_diesel/stock_line.rs [hunk 6]
```diff
@@ -160,15 +161,27 @@ impl<'a> StockLineRepository<'a> {
...
+                location::table,
+            >,
+            name::table,
+        >,
         barcode::table,
     >,
     DBType,
 >;
 
+allow_tables_to_appear_in_same_query!(stock_line, item_link);
```
Comment (roxy-dao) [inline server/repository/src/db_diesel/stock_line.rs:176]: Can these be moved to their respective row files?

Comment (Chris-Petty) [reply server/repository/src/db_diesel/stock_line.rs:176]: Sounds good!

Comment (Chris-Petty) [reply server/repository/src/db_diesel/stock_line.rs]: Done here 5f9ba3fe376ce8d81d436265440d28ab37c3be32

File: server/repository/src/db_diesel/item_row.rs [hunk 7]
```diff
@@ -1,6 +1,6 @@
 use super::{item_row::item::dsl::*, unit_row::unit, StorageConnection};
 
-use crate::repository_error::RepositoryError;
+use crate::{item_link, repository_error::RepositoryError};
```
Comment (Chris-Petty) [inline server/repository/src/db_diesel/item_row.rs:3]: In other files I've observed this import being stated as `item_link_row::item_link`, but this is what the rust-analyser suggested fix/syntax was so I just went with it 🙂. Is there a convention I'm missing?

Comment (roxy-dao) [reply server/repository/src/db_diesel/item_row.rs:3]: `item_link` doesn't have a another module could `item_link`, so it works... but for places that do, then it won't be importing the table...

File: server/repository/src/migrations/v1_06_00/mod.rs [hunk 8]
```diff
@@ -19,6 +21,8 @@ impl Migration for V1_06_00 {
         plugin_data::migrate(connection)?;
         master_list::migrate(connection)?;
         temperature_breach::migrate(connection)?;
+        item_link_create_table::migrate(connection)?;
+        item_line_add_item_link_id::migrate(connection)?;
```
Comment (roxy-dao) [inline server/repository/src/migrations/v1_06_00/mod.rs:25]: nit: I know mSupply names it this... but can this be renamed to stock_line_... 😄

File: server/repository/src/db_diesel/item_link_row.rs [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+            .on_conflict(id)
+            .do_update()
+            .set(item_link_row)
+            .execute(&self.connection.connection)?;
+        Ok(())
+    }
+
+    #[cfg(not(feature = "postgres"))]
+    pub fn upsert_one(&self, item_link_row: &ItemLinkRow) -> Result<(), RepositoryError> {
+        diesel::replace_into(item_link)
```
Comment (andreievg) [inline server/repository/src/db_diesel/item_link_row.rs:54]: Just a pesonal preference, I've been using, `use item_link:dsl as item_line_dsl`, and use that one rathern them item_link, just a bit more expicit about what time_line is, but is a lot more verbose. Also the overall import could just be use `super::{item_link_row::item_link::dsl:: ..``, -> `use self::item_link::dsl:: ..` (vs importing from super what's been exported here)

File: server/repository/src/db_diesel/item_link_row.rs [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        &self,
+        item_link_ids: &[String],
+    ) -> Result<Vec<ItemLinkRow>, RepositoryError> {
+        let result = item_link
+            .filter(id.eq_any(item_link_ids))
+            .load(&self.connection.connection)?;
+        Ok(result)
+    }
+
+    pub fn delete(&self, item_link_id: &str) -> Result<(), RepositoryError> {
```
Comment (andreievg) [inline server/repository/src/db_diesel/item_link_row.rs:93]: I would suspect we never delete item_links ?

**Review by mark-prins — CHANGES_REQUESTED** [review]
Changes please 🙏 
There's some migration script updates required for postgres to run - I've added suggestions for those.

The available quantity is no longer shown in the item list. that's coming from the item_stats which requires a linking from item.id to the stock_line.item_id. possibly fix in the `stock_on_hand` filter?

image

File: server/repository/src/migrations/v1_06_00/stock_line_add_item_link_id.rs [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        -- Recreate stock_on_hand taking into account new model
+        CREATE VIEW
+          store_items AS
+        SELECT
+          *
+        FROM
+          item
+          LEFT JOIN item_link ON item_link.item_id = item.id
+          LEFT JOIN stock_line ON stock_line.item_link_id = item_link.id
+          LEFT JOIN store ON store.id = stock_line.store_id;
```
[the remark this answers is not in the document]
Comment (Chris-Petty) [reply server/repository/src/migrations/v1_06_00/stock_line_add_item_link_id.rs:37]: Heh I was being lazy just returning `*` fields. The result in sqlite was interesting to say the least:

image

Seems it just adds a number for each joined table for each duplicate field, but gets wacky - `code` item and store, but gets column name `code:1` for the duplicate. But I'd have assumed store was table 3 in the query (with `id:3` for the column)... not worth understanding, better to specify columns properly haha.

**Comment by Chris-Petty** [comment]
> Changes please 🙏 There's some migration script updates required for postgres to run - I've added suggestions for those.
> 
> The available quantity is no longer shown in the item list. that's coming from the item_stats which requires a linking from item.id to the stock_line.item_id. possibly fix in the `stock_on_hand` filter?
> image

@mark-prins In my sqlite testing it seemed fine:

image

Before getting that image just now I was having something really odd happen though - it seemed that stock_lines would be lost during migration, then try reintegrate from buffer (left from initialisation) for some reason? I can't replicate it now though, either I was doing something subtly weird or there is a funky race condition? hmm.

Anyway, was yours not working because the SOH view was broken in postgres? OK now?

**Comment by andreievg** [comment]
I suggest another issue to fix postgres

File: server/repository/src/migrations/v1_06_00/item_line_add_item_link_id.rs [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        PRAGMA foreign_keys = OFF;
+        
+        ALTER TABLE stock_line
+        ADD COLUMN item_link_id TEXT NOT NULL REFERENCES item_link (id) DEFAULT 'temp_for_migration'; -- Can't have NOT NULL without a default... no PRAGMA for turning constraints off!
+        
+        UPDATE stock_line
+        SET
+        item_link_id = item_id;
+        
+        PRAGMA foreign_keys = ON;
```
[the remark this answers is not in the document]
Comment (Chris-Petty) [reply server/repository/src/migrations/v1_06_00/item_line_add_item_link_id.rs]: Ah yea, sqlite won't let you add constraints to columns after the fact 🙃. So have to do it ze old way.

**Comment by Chris-Petty** [comment]
image
Not to brag but I am a literal god

**Comment by andreievg** [comment]
Postgres migration changes look good to me, thanks.

I only ran the tests (sqlite and postgres), didn't do any manual testing

**Comment by Chris-Petty** [comment]
@mark-prins pls 🥺

File: server/repository/src/db_diesel/item_link_row.rs (hunk 9 above)
Comment (Chris-Petty) [reply server/repository/src/db_diesel/item_link_row.rs]: I agree decided to chuck in, thanks!

**Review by mark-prins — APPROVED** [review]

**Comment by mark-prins** [comment]
🙏  thanks Chris - took me ages to realise that I had to drop all the views and recreate things to get it working 🙄 
All good now

**Comment by Chris-Petty** [comment]
> 🙏 thanks Chris - took me ages to realise that I had to drop all the views and recreate things to get it working 🙄 All good now

Hmm was it not automatically working? I had a couple times where I confused myself where I was already on a 1.06 database so migrations didn't run heh. Is that what you mean?
